### ISSUE TITLE
Potential memory leak in the synchronous scanner

### ISSUE THREAD

**Issue by ericrange** [nabla-c0d3/sslyze#196]
if i run this in a endless loop, the memory is getting bigger and bigger. can anyone please help me. this is a really serious problem for me :(

sslv20 = { 'version': 'SSLv20', 'command': Sslv20ScanCommand() }
sslv30 = { 'version': 'SSLv30', 'command': Sslv30ScanCommand() }
tlsv10 = { 'version': 'TLSv10', 'command': Tlsv10ScanCommand() }
tlsv11 = { 'version': 'TLSv11', 'command': Tlsv11ScanCommand() }
tlsv12 = { 'version': 'TLSv12', 'command': Tlsv12ScanCommand() }

for protocol in [sslv20, sslv30, tlsv10, tlsv11, tlsv12]:
     scanner = SynchronousScanner(network_timeout=2, network_retries=2)
     scan_result = scanner.run_scan_command(server_info, protocol['command'])

...

**Comment by nabla-c0d3** [comment]
Does it still happen if you instantiate the `SynchronousScanner` outside of the loop? Do you have numbers? How long does it take and how much memory?

**Comment by ericrange** [comment]
yes it happens also there. well i have 6 threads... each threads got a infinite loop (6 ips).
it starts with ca 200mb and then grows after a whole loop, ca 50mb.
so i takes ca. 2 hours to fill 2gb ram.

**Comment by ericrange** [comment]
i tried https://pythonhosted.org/Pympler/muppy.html
and u can see there that lists objects increases drastically

**Comment by ericrange** [comment]
**for reproduce:**
from sslyze.synchronous_scanner import SynchronousScanner
from sslyze.plugins.openssl_cipher_suites_plugin import Tlsv12ScanCommand, Tlsv11ScanCommand, Tlsv10ScanCommand, Sslv30ScanCommand, Sslv20ScanCommand
from sslyze.server_connectivity import ServerConnectivityInfo, ServerConnectivityError

server_info = ServerConnectivityInfo(hostname='www.test.de, port=443)
server_info.test_connectivity_to_server()

sslv20 = { 'version': 'SSLv20', 'command': Sslv20ScanCommand() }
sslv30 = { 'version': 'SSLv30', 'command': Sslv30ScanCommand() }
tlsv10 = { 'version': 'TLSv10', 'command': Tlsv10ScanCommand() }
tlsv11 = { 'version': 'TLSv11', 'command': Tlsv11ScanCommand() }
tlsv12 = { 'version': 'TLSv12', 'command': Tlsv12ScanCommand() }

for protocol in [sslv20, sslv30, tlsv10, tlsv11, tlsv12]:
	scanner = SynchronousScanner(network_timeout=2, network_retries=2)
	scan_result = scanner.run_scan_command(server_info, protocol['command'])
	print(scan_result)

you can see that that every 3 seconds there is 1mb ram less.

**Comment by ericrange** [comment]
seems to be a nassl topic. the memory infinitly increases when i do:
```
#!/usr/bin/python2.7

import os
import sys
sys.path.insert(1, os.path.join(os.path.dirname(__file__), u'lib'))

from nassl.ssl_client import OpenSslVersionEnum
import socket
from nassl.debug_ssl_client import DebugSslClient


while True:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)
        sock.connect((u'test.com', 443))

        ssl_client = DebugSslClient(ssl_version=OpenSslVersionEnum.TLSV1_2, sock=sock, ssl_verify_locations=u'mozilla.pem')
        #ssl_client.set_tlsext_status_ocsp()
        ssl_client.do_handshake()

        print(u'Certificate chain')
        for cert in ssl_client.get_peer_cert_chain():
                print(cert.as_dict()[u'subject'][u'commonName'])

        print(u'\nCipher suite')
        print(ssl_client.get_current_cipher_name())

```

**Comment by ericrange** [comment]
is there a global var which never gets cleaned?

**Comment by ericrange** [comment]
try this:
you will see that each round adds each 2 list and str objects.

```
#!/usr/bin/python2.7

from pympler import muppy, summary

import os
import sys
sys.path.insert(1, os.path.join(os.path.dirname(__file__), u'lib'))

from nassl.ssl_client import OpenSslVersionEnum
import socket
from nassl.debug_ssl_client import DebugSslClient


while True:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)
        sock.connect((u'test.com', 443))

        ssl_client = DebugSslClient(ssl_version=OpenSslVersionEnum.TLSV1_2, sock=sock, ssl_verify_locations=u'mozilla.pem')
        #ssl_client.set_tlsext_status_ocsp()
        ssl_client.do_handshake()

        print(u'Certificate chain')
        for cert in ssl_client.get_peer_cert_chain():
                print(cert.as_dict()[u'subject'][u'commonName'])

        print(u'\nCipher suite')
        print(ssl_client.get_current_cipher_name())
        
        all_objects = muppy.get_objects()
        sum1 = summary.summarize(all_objects)
        summary.print_(sum1)
        all_objects = None
        sum1 = None
```

**Comment by ericrange** [comment]
sorry for all the spam. i just random remove some functions.
`cert.as_dict()` produces the memory leak.
but i guess that is not the only function.

**Comment by ericrange** [comment]
ok... seems there is no memory issue with the concurrent scanner :O
but still with sync scanner.

**Comment by nabla-c0d3** [comment]
The concurrent scanner uses multiple processes, so the memory leak may not be as obvious, but the code for scanning is exactly the same. I will get rid of `cert.as_dict()` in the next release (for other reasons) so that should help.

**Comment by ericrange** [comment]
ok. i just want to say, that (finally) the concurrent scanner do not produce any memory leaks.
i have been running my script now for 1 week and the memory is stable.
really strange that the sync scanner behaves so differently

**Comment by konklone** [comment]
I wanted to ring in and say I'm experiencing a very similar memory leak in 1.1.4 (current release). Definitely using the synchronous scanner, and I think the concurrent scanner as well (though it's more subtle, as you said -- I believe it generates stuck processes that way, whereas the synchronous scanner just gets fully killed in-process). I'm using pretty much the same scans @ericrange is (plus the cert info scanner).

I'll see if I can dig into it more, but it's definitely making batch scans challenging. Scanning ~20K hostnames on an EC2 server with 8GB of RAM consistently fails.

**Comment by konklone** [comment]
Actually -- I also can only reproduce the memory leak with the synchronous scanner. And it happens even if I don't use the certificate info plugin and only check for the main 5 protocol versions. But using a concurrent scanner, even over a large batch, doesn't demonstrate a memory leak.

**Comment by nabla-c0d3** [comment]
So just to recap, you are running the `—tlsv1` (and similar) commands against a very long list of host names, and the process gets killed when it runs out of memory?

**Comment by konklone** [comment]
@nabla-c0d3 Not quite, I'm using the `sslyze` library in Python, and initializing the scanner objects and executing/queueing commands there.

sslyze is invoked by 1 or more workers in a ThreadPoolExecutor. Each threaded worker can initialize either a SynchronousScanner or a ConcurrentScanner, depending on options passed to the worker orchestrator. (Yes, this is also a recipe for deadlocks, which is an issue I've been working on separately.)

Each threaded worker runs this code (sslyze-lib usage starting at about here):
https://github.com/18F/domain-scan/blob/master/scanners/sslyze.py#L137

I have noticed the memory leak even when only using 1 worker, but only when using the synchronous scanner.

**Comment by nabla-c0d3** [comment]
I did some testing to find memory leaks:

* When testing connectivity- https://gist.github.com/nabla-c0d3/4d08bf2cdebfef2b75aea493eab09817
* When using the synchronous scanner - https://gist.github.com/nabla-c0d3/fc633194f7a8509a7e68c931f1fbc471
* When doing an SSL handshake - https://gist.github.com/nabla-c0d3/02ffe4b6e34c5248a343f24af69ffcf3

None of these tests showed any obvious leaks so I am guessing it is happening in the C code, within nassl. I opened an issue there: https://github.com/nabla-c0d3/nassl/issues/21

**Comment by nabla-c0d3** [comment]
I fixed some big memory leaks in nassl - the fixes have been released as part of nassl 1.0.2. Memory usage should be a lot better now.

**Comment by konklone** [comment]
Thank you for tracking these down!
